internal/dag/builder_test.go

```diff
@@ -6056,6 +6058,85 @@ func TestSplitSecret(t *testing.T) {
 	}
 }
 
+func TestValidateHeaderAlteration(t *testing.T) {
+	tests := []struct {
+		name       string
+		in         *projcontour.HeaderAlterations
+		wantAdd    map[string]string
+		wantRemove []string
+		wantErr    error
+	}{{
+		name: "empty is fine",
+	}, {
+		name: "add two, remove one",
+		in: &projcontour.HeaderAlterations{
+			Add: []projcontour.HeaderAddition{{
+				Name:  "K-Foo",
+				Value: "bar",
+			}, {
+				Name:  "k-baz", // This gets canonicalized
+				Value: "blah",
+			}},
+			Remove: []string{"K-Nada"},
+		},
+		wantAdd: map[string]string{
+			"K-Foo": "bar",
+			"K-Baz": "blah",
+		},
+		wantRemove: []string{"K-Nada"},
+	}, {
+		name: "duplicate add",
+		in: &projcontour.HeaderAlterations{
+			Add: []projcontour.HeaderAddition{{
+				Name:  "K-Foo",
+				Value: "bar",
+			}, {
+				Name:  "k-foo", // This gets canonicalized
+				Value: "blah",
+			}},
+		},
+		wantErr: errors.New(`Duplicate header addition: "K-Foo"`),
+	}, {
+		name: "duplicate remove",
+		in: &projcontour.HeaderAlterations{
+			Remove: []string{"K-Foo", "k-foo"},
+		},
+		wantErr: errors.New(`Duplicate header removal: "K-Foo"`),
+	}, {
+		name: "invalid add header",
+		in: &projcontour.HeaderAlterations{
+			Add: []projcontour.HeaderAddition{{
+				Name:  "  K-Foo",
+				Value: "bar",
+			}},
+		},
+		wantErr: errors.New(`invalid add header "  K-Foo": [a valid HTTP header must consist of alphanumeric characters or '-' (e.g. 'X-Header-Name', regex used for validation is '[-A-Za-z0-9]+')]`),
```

Comment:
Is this actually the regex api machinery uses for validation? It's way too restrictive for HTTP

https://httpwg.org/specs/rfc7230.html#header.fields